### PR TITLE
Make FailingSnapshotManager Thread Safe

### PR DESCRIPTION
**Pull request by gabber235**
# Problem Description
When running multiple path searches at the same time from different threads, the `FailingSnapshotManager` throws the following error:

```
java.util.ConcurrentModificationException: null
	at java.util.HashMap.computeIfAbsent(HashMap.java:1221) ~[?:?]
	at org.patheloper.model.snapshot.FailingSnapshotManager.processChunkSnapshot(FailingSnapshotManager.java:108) ~[Typewriter.jar:?]
	at org.patheloper.model.snapshot.FailingSnapshotManager.getChunkSnapshot(FailingSnapshotManager.java:96) ~[Typewriter.jar:?]
	at org.patheloper.model.snapshot.FailingSnapshotManager.fetchBlock(FailingSnapshotManager.java:59) ~[Typewriter.jar:?]
	at org.patheloper.model.snapshot.FailingSnapshotManager.getBlock(FailingSnapshotManager.java:116) ~[Typewriter.jar:?]
	at org.patheloper.model.snapshot.FailingSnapshotManager$RequestingSnapshotManager.getBlock(FailingSnapshotManager.java:168) ~[Typewriter.jar:?]
	at org.patheloper.model.pathing.pathfinder.AbstractPathfinder.isBlockUnreachable(AbstractPathfinder.java:112) ~[Typewriter.jar:?]
	at org.patheloper.model.pathing.pathfinder.AbstractPathfinder.initialChecksFailed(AbstractPathfinder.java:104) ~[Typewriter.jar:?]
	at org.patheloper.model.pathing.pathfinder.AbstractPathfinder.findPath(AbstractPathfinder.java:71) ~[Typewriter.jar:?]
	at org.patheloper.model.pathing.pathfinder.AStarPathfinder.findPath(AStarPathfinder.java:34) ~[Typewriter.jar:?]
```

Proposed solution is to change the data structure of the `SNAPSHOTS_MAP` from a unsafe `Map` to a thread safe `ConcurrentMap`.

This pull request does just that.

### CODE REVIEW

File: pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
  */
 public class FailingSnapshotManager implements SnapshotManager {
 
-  private static final Map<UUID, WorldDomain> SNAPSHOTS_MAP = new HashMap<>();
+  private static final ConcurrentMap<UUID, WorldDomain> SNAPSHOTS_MAP = new ConcurrentHashMap<>();
```
Comment (olijeffers0n) [inline pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java]: No need to store this as a ConcurrentMap, we can keep this as a Map here.

**Review by olijeffers0n — CHANGES_REQUESTED** [review]
Hi, thanks for this! Could you target our trunk branch with this PR please

File: pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java (hunk 1 above)
Comment (gabber235) [reply pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java:40]: Another way to allow for multi threaded access is to use a Mutex. But that is more complicated. Instead the concurrent map just takes care of it and doesn’t change the api

Comment (gabber235) [reply pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java]: Done

**Review by olijeffers0n — APPROVED** [review]
LGTM - @GodCipher can final approval and merge

File: pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java [hunk 2]
```diff
@@ -21,6 +20,8 @@
 import org.patheloper.util.ChunkUtils;
 import org.patheloper.util.ErrorLogger;
 
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
```
Comment (bsommerfeld) [inline pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java]: Please remove this unused import

**Review by bsommerfeld — CHANGES_REQUESTED** [review]
Hi, thanks for the fix tho! Once the import is removed we will merge this into pathetic :)

File: pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java (hunk 2 above)
Comment (gabber235) [reply pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java]: Fixed

**Review by bsommerfeld — APPROVED** [review]

**Pull request merged**